framework/src/controller/application_state.js

```diff
@@ -0,0 +1,110 @@
+/* eslint-disable class-methods-use-this */
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+'use strict';
+
+const os = require('os');
+const crypto = require('crypto');
+const _ = require('lodash');
+
+/**
+ * Initial state of the entire application.
+ * - os
+ * - version
+ * - wsPort
+ * - httpPort
+ * - minVersion
+ * - protocolVersion
+ * - height
+ * - nethash
+ * - broadhash
+ * - nonce
+ *
+ * @class
+ * @requires crypto
+ * @requires os
+ * @param {Object} initialState - Initial state of the application
+ * @param {Object} logger
+ */
+class ApplicationState {
+	constructor({
+		initialState: {
+			version,
+			wsPort,
+			httpPort,
+			minVersion,
+			protocolVersion,
+			nethash,
+			nonce,
+		},
+		logger,
+	}) {
+		this.logger = logger;
+		this.state = {
+			os: os.platform() + os.release(),
+			version,
+			wsPort,
+			httpPort,
+			minVersion,
+			protocolVersion,
+			height: 1,
+			nethash,
+			broadhash: nethash,
+			nonce,
+		};
```

Comment:
When you assign `state` this `this` then its accessible and modifiable without calling the `update`. So it would be nice to use the `WeakMap` approach that we used actually have private data members. 

Then you can simply have, 

```js
get state() {
    return _.cloneDeep(stateWeakMap.get(this));
}
```

So now one can't miss to change the state without performing full update logic.